thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/datasource/cache/metadata/MetaDataCollectionCache.java

```diff
@@ -0,0 +1,91 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ *
+ */
+
+package org.apache.pinot.thirdeye.datasource.cache.metadata;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The cache for a meta-data collection.
+ * This cache will hold a collection of meta-data and periodically reloads it using the cache loader.
+ *
+ * It's designed to cache the original or derived meta-data collection for ThirdEye Dashboard server, for example, a
+ * list of all applications, subscription groups, and detection configs. Loading a collection of metadata has been
+ * a pain point for ThirdEye. It requires data base to scan the index table and then do a look up for each config in the
+ * generic json table, which is slow. This cache makes such loading asynchronous and returns user the cached result
+ * when opening page.
+ *
+ * @param <T> the type of the value to be cached.
+ */
+public class MetaDataCollectionCache<T> {
+  private static final String META_DATA_KEY = "metaData";
+
+  private final LoadingCache<String, T> loadingCache;
+  private final ExecutorService adHocRefreshExecutor;
+  /**
+   * Create a Metadata collection cache
+   * @param cacheLoader the cache loader for the meta data collection
+   * @param reloadPeriod the period of the reloading frequency
+   * @param reloadUnit the unit for the reload period
+   */
+  public MetaDataCollectionCache(MetadataCacheLoader<T> cacheLoader, int reloadPeriod, TimeUnit reloadUnit) {
+    this.loadingCache = CacheBuilder.newBuilder()
+        .expireAfterWrite(10, TimeUnit.MINUTES)
+        .build(CacheLoader.from(cacheLoader::loadMetaData));
+
+    ScheduledExecutorService scheduledCacheRefreshExecutor = Executors.newSingleThreadScheduledExecutor();
+    // refresh the cached value periodically. This will also load the data into cache when the
+    // application starts.
+    scheduledCacheRefreshExecutor.scheduleAtFixedRate(this::refresh, 0, reloadPeriod, reloadUnit);
+    adHocRefreshExecutor = Executors.newSingleThreadExecutor();
+  }
+
+  /**
+   * Create a Metadata collection cache. By default reloads every 3 minutes
+   * @param cacheLoader the cache loader for the meta data collection
+   */
+  public MetaDataCollectionCache(MetadataCacheLoader<T> cacheLoader) {
+    this(cacheLoader, 3, TimeUnit.MINUTES);
```

Comment:
Would this offer more load on mysql? In our current model, we query the db only when users load the alerts page which is very rare. Refreshing every 3 minutes would result in a lot of unnecessary queries to mysql right.

Reply:
Yes, it will query the DB more times. For our current usage pattern, this is called around 100 times a day, but mostly during the work hours. Keeping the cache fresh will require more queries, but will be more spread out to the whole day. The QPS doesn't vary much. As the number of the page view grows, it will benefit us more.

Reply:
Keep cache consistent is a potential issue as @akshayrai pointed out. Actually mysql has cache itself and I don't think this is a big issue for now.

Reply:
okay, I'll remove it for now